### PR TITLE
Create synthetic `exports` symbol for commonjs `module`

### LINKED ISSUE #41764: In JS, checking `require.main === module` causes the type of `module` to get messed up
**TypeScript Version:**  4.1.2

**Search Terms:** require.main module

When using Node.js's [recommended approach](https://github.com/nodejs/node/blob/master/doc/api/deprecations.md#dep0144-moduleparent) to check if a module is the main module, tsc chokes on the types it infers:

**Code**
**Note:** This is JavaScript!
```js
if (require.main === module) { // ERROR TS2367 The types "Module | undefined" and "{ "\"c:/Repositories/ioBroker.test/repro\"": number; }" have no overlap
	module.exports = 1; // ERROR TS2339 The property exports does not exist on type `never`
}
```

**Expected behavior:**
no error

**Actual behavior:**
bunch of incorrect errors (see code). I'm not even sure what that type which is inferred for `module` is supposed to mean.

**Playground Link:** doesn't work on the playground

**Related Issues:** https://github.com/microsoft/TypeScript/issues/30088

### PR DESCRIPTION
**Pull request by sandersn**
Previously, the `module` identifier in commonjs modules got a synthetic type with a single property `exports`. The exports property reused the file's symbol, which, for a module file, gives the correct exported properties.

However, the name of this symbol was still the filename of the file, not `exports`. This PR creates a synthetic symbol for `exports` by copying in a similar way to esModuleInterop's `default` symbol in `resolveESModuleSymbol` (although the intent there is to strip off signatures from the symbol).

It's not possible to use `cloneSymbol` because that function merges the symbols, altering the original to point to the new symbol.

Fixes #41764

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8768,8 +8768,15 @@ namespace ts {
             }
             if (symbol.flags & SymbolFlags.ModuleExports) {
                 const fileSymbol = getSymbolOfNode(getSourceFileOfNode(symbol.valueDeclaration));
+                const result = createSymbol(fileSymbol.flags, "exports" as __String);
+                result.declarations = fileSymbol.declarations ? fileSymbol.declarations.slice() : [];
+                result.parent = fileSymbol.parent;
```
Comment (weswigham) [inline src/compiler/checker.ts:8773]: The `parent` should be reset to the `module` symbol in `module.exports`, no?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8768,8 +8768,15 @@ namespace ts {
             }
             if (symbol.flags & SymbolFlags.ModuleExports) {
                 const fileSymbol = getSymbolOfNode(getSourceFileOfNode(symbol.valueDeclaration));
+                const result = createSymbol(fileSymbol.flags, "exports" as __String);
+                result.declarations = fileSymbol.declarations ? fileSymbol.declarations.slice() : [];
+                result.parent = symbol;
+                result.target = fileSymbol;
+                if (fileSymbol.valueDeclaration) result.valueDeclaration = fileSymbol.valueDeclaration;
+                if (fileSymbol.members) result.members = new Map(fileSymbol.members);
+                if (fileSymbol.exports) result.exports = new Map(fileSymbol.exports);
```
Comment (andrewbranch) [inline src/compiler/checker.ts:8777]: How much of a memory hit for big checkJs projects (that use `module.exports`) is this?

**Comment by sandersn** [comment]
@typescript-bot user test this

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [reply src/compiler/checker.ts:8777]: The only baseline we have for this is "does chrome-devtools-frontend run out of memory?". I'll run the user tests.

A priori, it shouldn't be gigantic; a file with 1000 exports will add 1000 map entries. The first page of google only returns [an SO answer from a V8 developer](https://stackoverflow.com/a/54466812/7851), which points out that each entry has key,value and next pointers. So...is that 6000 bytes per large file? It's not too expensive, but on the other hand this is a mostly cosmetic feature. The exception is the bug that the PR fixes.

**Review by andrewbranch — APPROVED** [review]

**Comment by sandersn** [comment]
User test baselines finished OK, so I think this is safe enough.
(It also found a couple instances of the pattern from the original bug.)

**Pull request merged**